### PR TITLE
Include BCL assemblies required for VS 2017 build tools

### LINKED ISSUE #589: FileNotFoundException for System.Memory when using Visual Studio Build Tools 2017
I tried to upgrade Nerdbank.GitVersioning from 3.3.37 to 3.4.194 but got a System.IO.FileNotFoundException for the System.Memory assembly. This is with MSBuild.exe ("Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework") installed as part of Visual Studio Build Tools 2017 version 15.9.34, on Windows Server 2012 R2.

```
Project "C:\REDACTED.csproj.metaproj" (10:42) is building "C:\REDACTED.csproj" (11:9) on node 1 (default targets).
Project "C:\REDACTED.csproj" (11:9) is building "C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets" (127) on node 1 (GetBuildVersion_Properties target(s)).
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: The "Nerdbank.GitVersioning.Tasks.GetBuildVersion" task failed unexpectedly.
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: System.IO.FileNotFoundException: Could not load file or assembly 'System.Memory, Version=4.0.1.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified.
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: File name: 'System.Memory, Version=4.0.1.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitRepository..ctor(String workingDirectory, String gitDirectory, String commonDirectory, String objectDirectory)
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitRepository.Create(String workingDirectory)
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.ManagedGitContext..ctor(String workingDirectory, String dotGitPath, String committish)
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.GitContext.Create(String path, String committish, Boolean writable)
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Tasks.GetBuildVersion.ExecuteInner()
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at MSBuildExtensionTask.ContextAwareTask.Execute()
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: 
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: WRN: Assembly binding logging is turned OFF.
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: Note: There is some performance penalty associated with assembly bind failure logging.
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].
C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: 
Done Building Project "C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\Nerdbank.GitVersioning.Inner.targets" (GetBuildVersion_Properties target(s)) -- FAILED.
Done Building Project "C:\REDACTED.csproj" (default targets) -- FAILED.
Project "C:\REDACTED.csproj.metaproj" (10:42) is building "C:\REDACTED.csproj" (5:81) on node 1 (default targets).
```

Process Monitor shows it loads `C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\MSBuildFull\Nerdbank.GitVersioning.Tasks.dll` and not `C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\MSBuildCore\Nerdbank.GitVersioning.Tasks.dll`. That seems correct to me.

"C:\Users\REDACTED\.nuget\packages\nerdbank.gitversioning\3.4.194\build\MSBuildFull\NerdBank.GitVersioning.dll" references the missing assembly:

```
.assembly extern System.Memory
{
  .publickeytoken = (CC 7B 13 FF CD 2D DD 51 )                         // .{...-.Q
  .ver 4:0:1:1
}
```

### PR DESCRIPTION
**Pull request by AArnott**
Include BCL assemblies required for VS 2017 build tools

Several assemblies are required by this msbuild task but are not included in MSBuild 2017. 

Adding the assemblies to the package alone is not enough, since binding redirects are also missing in msbuild.exe.config. We make up for this with a custom assembly resolver.

Fixes #589

### CODE REVIEW

**Comment by AArnott** [comment]
@KalleOlaviNiemitalo  could you please verify that this addresses the problem? You can download [the package built by this PR](https://dev.azure.com/andrewarnott/OSS/_build/results?buildId=4552&view=artifacts&pathAsName=false&type=publishedArtifacts) for testing from the `deployables` artifact.

**Comment by KalleOlaviNiemitalo** [comment]
@AArnott thank you. I don't know whether I'll have time to test it today.  Before Monday for sure.

**Comment by AArnott** [comment]
I tried testing it myself, but it seems SDK style projects can't build with just the VS 2017 Build Tools installed. I filed a bug against them on that in the meantime.

**Comment by AArnott** [comment]
I found a way to test it. It doesn't work because another assembly is then missing. I'm working on a better fix.

File: src/Nerdbank.GitVersioning.Tasks/Nerdbank.GitVersioning.nuspec [hunk 1]
```diff
@@ -27,6 +27,10 @@ IMPORTANT: The 3.x release may produce a different version height than prior maj
     <file src="$BaseOutputPath$net461\Validation.dll" target="build\MSBuildFull\Validation.dll" />
     <file src="$BaseOutputPath$net461\System.Runtime.CompilerServices.Unsafe.dll" target="build\MSBuildFull\System.Runtime.CompilerServices.Unsafe.dll" />
     <file src="$BaseOutputPath$net461\PInvoke.Kernel32.dll" target="build\MSBuildFull\PInvoke.Kernel32.dll" />
+    <file src="$BaseOutputPath$net461\System.Buffers.dll" target="build\MSBuildFull\System.Buffers.dll" />
+    <file src="$BaseOutputPath$net461\System.Memory.dll" target="build\MSBuildFull\System.Memory.dll" />
+    <file src="$BaseOutputPath$net461\System.Runtime.CompilerServices.Unsafe.dll" target="build\MSBuildFull\System.Runtime.CompilerServices.Unsafe.dll" />
```
Comment (KalleOlaviNiemitalo) [inline src/Nerdbank.GitVersioning.Tasks/Nerdbank.GitVersioning.nuspec:32]: It looks like you listed System.Runtime.CompilerServices.Unsafe twice.

**Comment by KalleOlaviNiemitalo** [comment]
<https://dev.azure.com/andrewarnott/OSS/_build/results?buildId=4555&view=artifacts&pathAsName=false&type=publishedArtifacts> (#3.4.196+e22f3b6a14) seems to be the build with the latest changes.

**Comment by AArnott** [comment]
You can get it now per the instructions on the README from our CI feed, if you want it before we push to nuget.org.

**Pull request merged**